lisa/operating_system.py

```diff
@@ -390,7 +390,17 @@ def install_packages(
         extra_args: Optional[List[str]] = None,
     ) -> None:
         package_names = self._get_package_list(packages)
-        self._install_packages(package_names, signed, timeout, extra_args)
+        max_retries = 3
+        for attempt in range(1, max_retries + 1):
+            try:
+                self._install_packages(package_names, signed, timeout, extra_args)
+                break
+            except Exception as e:
+                if attempt == max_retries:
+                    raise  # re-raise after final attempt
+                else:
+                    print(f"install_packages attempt {attempt} failed: {e}, retrying")
```

Comment:
Don't use `print`, use `log.debug`. Always using it when developing, so it won't be merged incidentally.